compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java

```diff
@@ -491,7 +514,33 @@ static void generateCodeInternal(ModuleContext moduleContext,
         // Write the bir to the file system
         // This code will execute only if JAR caching is successful
         // TODO: check the filesystem cache and delete if the cache is incomplete (if BIR or JAR is missing)
-        moduleContext.compilationCache.cacheBir(moduleContext.moduleName(), birContent);
+        if (shouldNotOptimizeModule) {
+            moduleContext.compilationCache.cacheBir(moduleContext.moduleName(), birContent);
+        }
+    }
+
+    public boolean isWhiteListedModule() {
+        return isObserveModule() || isDriverModule() || isJBallerinaModule();
+    }
+
+    private boolean isObserveModule() {
+        return this.moduleId.moduleName().contains(OBSERVE);
```

Comment:
this `moduleName` function is deprecated. Better to use another approach.

Reply:
Fixed in aa8254a61e2f3d88c1fdd944ffe5a5b257ede583.